src/mscorlib/shared/System/Buffers/OwnedMemory.cs

```diff
@@ -0,0 +1,53 @@
+// Licensed to the .NET Foundation under one or more agreements.
+// The .NET Foundation licenses this file to you under the MIT license.
+// See the LICENSE file in the project root for more information.
+
+using System.Runtime;
+using System.Runtime.CompilerServices;
+
+namespace System.Buffers
+{
+    public abstract class OwnedMemory<T> : IDisposable, IRetainable
+    {
+        public abstract int Length { get; }
+
+        public abstract Span<T> AsSpan(int index, int length);
+
+        public virtual Span<T> AsSpan()
+        {
+            if (IsDisposed) ThrowHelper.ThrowObjectDisposedException(nameof(OwnedMemory<T>), ExceptionResource.Memory_ThrowIfDisposed);
+            return AsSpan(0, Length);
+        }
+
+        public Memory<T> AsMemory
+        {
+            get 
+            {
+                if (IsDisposed) ThrowHelper.ThrowObjectDisposedException(nameof(OwnedMemory<T>), ExceptionResource.Memory_ThrowIfDisposed);
+                return new Memory<T>(this, 0, Length);
+            }
+        }
+
+        public abstract MemoryHandle Pin();
+
+        protected internal abstract bool TryGetArray(out ArraySegment<T> arraySegment);
+
+        public void Dispose()
+        {
+            if (IsRetained) ThrowHelper.ThrowInvalidOperationException(ExceptionResource.Memory_OutstandingReferences);
```

Comment:
:bulb: These one-line statements are particularly difficult to read.

Reply:
❗️ I'm not a fan of the interaction between `IRetainable` and `IDisposable` for this type. I would expect the `Dispose` semantics to be equivalent to a case where the constructor called `Retain` and then `Dispose` calls release, except with additional handling to ensure `Dispose` is idempotent.

Reply:
We want to enable types with reference counting which means that a concrete OwnedMemory can call retain/release multiple times before disposing.

Reply:
:memo: This comment is related to my suggestion to review the semantics of `ReferenceCountedDisposable<T>` just to see if it's possible to unify the concepts.